### PR TITLE
Add possibility for custom event layouts

### PR DESCRIPTION
**Pull request by FHellmann**
see #3

### CODE REVIEW

**Comment by Tibolte** [comment]
Thanks for the contribution @FHellmann !

I tested a bit and I think first thing you need to check is this crash:

java.lang.NullPointerException
                                                                               at com.github.tibolte.sample.DrawableEventRenderer.render(DrawableEventRenderer.java:21)
                                                                               at com.github.tibolte.sample.DrawableEventRenderer.render(DrawableEventRenderer.java:10)
                                                                               at com.github.tibolte.agendacalendarview.agenda.AgendaAdapter.getView(AgendaAdapter.java:100)
                                                                               at se.emilsjolander.stickylistheaders.AdapterWrapper.getView(AdapterWrapper.java:164)
                                                                               at se.emilsjolander.stickylistheaders.AdapterWrapper.getView(AdapterWrapper.java:26)
                                                                               at android.widget.AbsListView.obtainView(AbsListView.java:2161)
                                                                               at android.widget.ListView.makeAndAddView(ListView.java:1840)
                                                                               at android.widget.ListView.fillUp(ListView.java:709)
                                                                               at android.widget.ListView.fillGap(ListView.java:648)
                                                                               at android.widget.AbsListView.trackMotionScroll(AbsListView.java:4970)
                                                                               at android.widget.AbsListView$FlingRunnable.run(AbsListView.java:4127)
                                                                               at android.view.Choreographer$CallbackRecord.run(Choreographer.java:749)
                                                                               at android.view.Choreographer.doCallbacks(Choreographer.java:562)
                                                                               at android.view.Choreographer.doFrame(Choreographer.java:531)
                                                                               at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:735)
                                                                               at android.os.Handler.handleCallback(Handler.java:730)
                                                                               at android.os.Handler.dispatchMessage(Handler.java:92)
                                                                               at android.os.Looper.loop(Looper.java:137)
                                                                               at android.app.ActivityThread.main(ActivityThread.java:5103)
                                                                               at java.lang.reflect.Method.invokeNative(Native Method)
                                                                               at java.lang.reflect.Method.invoke(Method.java:525)
                                                                               at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:737)
                                                                               at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
                                                                               at dalvik.system.NativeStart.main(Native Method)

**Comment by FHellmann** [comment]
The NullPointerException should be fixed with my latest commit.

File: app/src/main/res/layout/view_agenda_drawable_event.xml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+              android:paddingLeft="15dp"
+              android:paddingRight="15dp"
+              android:paddingTop="5dp">
+
+    <!-- Description -->
+    <LinearLayout
+            android:id="@+id/view_agenda_event_description_container"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginLeft="55dp"
```
Comment (Tibolte) [inline app/src/main/res/layout/view_agenda_drawable_event.xml]: We'd need to make that margin automatic, for instance in AgendaEventView's layout. The developer shouldn't have to do that, see comment from @chaossss here: https://github.com/Tibolte/AgendaCalendarView/issues/3

Comment (FHellmann) [inline app/src/main/res/layout/view_agenda_drawable_event.xml]: I've also thought about that. I will have a look at it.

Comment (FHellmann) [inline app/src/main/res/layout/view_agenda_drawable_event.xml]: Added with commit [8fe1fd3](https://github.com/FHellmann/AgendaCalendarView/commit/8fe1fd3fc4be6f3da18906249f9b9eba47a42b2c).

File: agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/agenda/AgendaAdapter.java [hunk 2]
```diff
@@ -76,12 +82,24 @@ public long getItemId(int position) {
 
     @Override
     public View getView(int position, View convertView, ViewGroup parent) {
```
Comment (Tibolte) [inline agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/agenda/AgendaAdapter.java]: I know I didn't do it before but now I'm realizing that the recycling of the views are not optimized. Looking at: https://github.com/emilsjolander/StickyListHeaders

Here is how it's done there in the example:

 @Override 
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder holder;

        if (convertView == null) {
            holder = new ViewHolder();
            convertView = inflater.inflate(R.layout.test_list_item_layout, parent, false);
            holder.text = (TextView) convertView.findViewById(R.id.text);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

        holder.text.setText(countries[position]);

        return convertView;
    }

Is it possible to apply this recycling pattern with your logic here?

File: agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/agenda/AgendaEventView.java [hunk 3]
```diff
@@ -1,69 +1,38 @@
...
 
-    public AgendaEventView(Context context) {
-        super(context);
+    public AgendaEventView(Context context, AttributeSet attrs) {
+        this(context, attrs, 0);
     }
 
     public AgendaEventView(Context context, AttributeSet attrs, int defStyleAttr) {
         super(context, attrs, defStyleAttr);
+        setPadding(getResources().getDimensionPixelSize(R.dimen.agenda_event_view_padding_left),
```
Comment (Tibolte) [inline agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/agenda/AgendaEventView.java]: :+1:

File: agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/render/EventRenderer.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.github.tibolte.agendacalendarview.render;
+
+import android.support.annotation.LayoutRes;
+import android.view.View;
+
+import com.github.tibolte.agendacalendarview.models.CalendarEvent;
+
+import java.lang.reflect.ParameterizedType;
+
+public abstract class EventRenderer<T extends CalendarEvent> {
```
Comment (Tibolte) [inline agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/render/EventRenderer.java]: Could you comment this class?

File: app/src/main/java/com/github/tibolte/sample/DrawableCalendarEvent.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
+package com.github.tibolte.sample;
+
+import com.github.tibolte.agendacalendarview.models.BaseCalendarEvent;
+import com.github.tibolte.agendacalendarview.models.CalendarEvent;
+
+import java.util.Calendar;
+
+public class DrawableCalendarEvent extends BaseCalendarEvent {
```
Comment (Tibolte) [inline app/src/main/java/com/github/tibolte/sample/DrawableCalendarEvent.java]: Could you comment this one also?

**Comment by Tibolte** [comment]
Thanks for your contribution @FHellmann :+1:

**Comment by FHellmann** [comment]
It was a pleasure to help you.

I haven't had the time to refactor the code more specificly. But I have no idea how to refactor it. The inspiration is still missing.

When I will find the time and inspriation, I will go for it and create a new pull request.

**Comment by iamdeadman** [comment]
So, I was also trying to use custom layouts for the events in the agenda list. 

For this I tried to define a custom event renderer class like this - 

```
public class CustomEventRenderer extends EventRenderer<BaseCalendarEvent> {

    @Override
    public void render(@NonNull View view, @NonNull BaseCalendarEvent event) {
        TextView txtTitle = (TextView) view.findViewById(R.id.view_agenda_event_title);
        TextView txtDescription = (TextView) view.findViewById(R.id.view_agenda_event_description);
        TextView txtLocation = (TextView) view.findViewById(R.id.view_agenda_event_location);
        LinearLayout descriptionContainer = (LinearLayout) view.findViewById(R.id.view_agenda_event_description_container);
        LinearLayout locationContainer = (LinearLayout) view.findViewById(R.id.view_agenda_event_location_container);

        try {
            descriptionContainer.setVisibility(View.VISIBLE);
            txtTitle.setTextColor(view.getResources().getColor(R.color.black));
            descriptionContainer.setBackgroundColor(event.getColor());
            txtLocation.setTextColor(view.getResources().getColor(com.github.tibolte.agendacalendarview.R.color.theme_text_icons));
            if (event.getTitle().equals(view.getResources().getString(com.github.tibolte.agendacalendarview.R.string.agenda_event_no_events))) {
                txtTitle.setTextColor(view.getResources().getColor(android.R.color.black));
            } else {
                txtTitle.setTextColor(view.getResources().getColor(com.github.tibolte.agendacalendarview.R.color.theme_text_icons));
            }
        } catch (Resources.NotFoundException ignored) {}

        try {
            if(event.getDescription() != null) txtDescription.setText(event.getDescription());
            if (event.getLocation() != null && event.getLocation().length() > 0) {
                locationContainer.setVisibility(View.VISIBLE);
                txtLocation.setText(event.getLocation());
            } else {
                locationContainer.setVisibility(View.GONE);
            }

            if(event.getTitle() != null) {
                txtTitle.setText(event.getTitle());
            }
        } catch (Exception ignored) {}
    }

    @Override
    public int getEventLayout() {
        return R.layout.custom_agenda_event;
    }
}

```

I am adding this to my view with - 
mAgendaCalendarView.init(eventList, minDate, maxDate, Locale.ENGLISH, this);
mAgendaCalendarView.addEventRenderer(new CustomEventRenderer());
            
Now in the AgendaAdapter I see the following function - 

```
@Override
    public View getView(int position, View convertView, ViewGroup parent) {
        EventRenderer eventRenderer = new DefaultEventRenderer();
        final CalendarEvent event = getItem(position);

        // Search for the correct event renderer
        for (EventRenderer renderer : mRenderers) {
            if(event.getClass().isAssignableFrom(renderer.getRenderType())) {
                eventRenderer = renderer;
                break;
            }
        }
        convertView = LayoutInflater.from(parent.getContext())
                .inflate(eventRenderer.getEventLayout(), parent, false);
        eventRenderer.render(convertView, event);
        return convertView;
    }
```

So, it seems that only the DefaultEventRenderer is getting inflated but not the customeventrenderer. 

Any suggestions as to how to make it work ?

**Comment by FHellmann** [comment]
Hi @iamdeadman,
the DefaultEventRenderer is called because your renderer also is for the BaseCalendarEvent model. Create your own model and "register" it in the generic type in your renderer class. If this still shouldn't work, override the getRenderType method in your renderer.

**Comment by iamdeadman** [comment]
Oh i see I forgot to override the BaseCalendarView model class. 
Didn't need to override the getRenderType as It now works after using a custom class with both event renderer and event model.  
thanks for the help!

**Comment by iamdeadman** [comment]
@FHellmann can we also override the day labels in the calendar view in a similar way ? 
Right now we have the timeline view with events in it but there are no labels on the day names corresponding to those events in the timeline so it is difficult for users to navigate through the entire timeline. 
Which classes should we override in order to achieve this ? 
Also can the floating action button be hidden or its functions can be overriden ? if so can you please provide a reference as to how this can be achieved.

**Comment by FHellmann** [comment]
@iamdeadman I'm not quite sure what you mean by "day labels"? Could you may provide a screenshot of what you mean and how it should look like?
The floating action button behaviour is set in the CalendarAgendaView with a listener for the scroll detection. You can register your own listener to the view and use the enableFloatingIndicator(boolean) method of CalendarAgendaView to show/hide the floating action button.

**Comment by iamdeadman** [comment]
Sorry I meant the date labels in the calendar both when it is in expanded and collapsed mode, Screenshot attached below - 
screenshot_2018-09-20-19-31-13

And for the floating labels I can see only these methods available in my IDE with the calendarview object - 

image

I am using this version in my build.gradle project file - 

compile 'com.github.tibolte:agendacalendarview:1.0.4'

and using butterknife to bind the view like this - 

@BindView(R.id.agenda_calendar_view)
AgendaCalendarView mAgendaCalendarView;

So, where should I look for this method - enableFloatingIndicator ?

**Comment by FHellmann** [comment]
I'm sorry, but there is no feature implemented to customize the day labels. If you need something like this, feel free to create a pull request.

The method should be available if you call [mAgendaCalendarView.enableFloatingIndicator(boolean)](https://github.com/Tibolte/AgendaCalendarView/blob/master/agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/AgendaCalendarView.java#L248).

**Comment by iamdeadman** [comment]
Okay, If you can spare time, can you please suggest where can I find the existing code that creates these date labels then I will start working on the PR for that. 

In the mean time, is there an option to remove / hide empty event dates from the agenda timeline (we only want to show dates which have any event associated with them) so if not I will try to add this feature too to this PR

**Comment by FHellmann** [comment]
The date labels for the timeline view are created in the [AgendaHeaderView](https://github.com/Tibolte/AgendaCalendarView/blob/master/agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/agenda/AgendaHeaderView.java). The date labels in the calendar are created in the [CalendarView](https://github.com/Tibolte/AgendaCalendarView/blob/master/agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/calendar/CalendarView.java) and [WeeksAdapter](https://github.com/Tibolte/AgendaCalendarView/blob/master/agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/calendar/weekslist/WeeksAdapter.java) for the small calendar view.

There is still no function to hide dates in the timeline when no event is available. You have to implement this feature in the [AgendaAdapter](https://github.com/Tibolte/AgendaCalendarView/blob/master/agendacalendarview/src/main/java/com/github/tibolte/agendacalendarview/agenda/AgendaAdapter.java).

**Comment by iamdeadman** [comment]
So, I cloned the repo and Gradle in my IntelliJ Android Studio was trying to sync the projects but I am getting this build error because of which my build is failing -  

No such property: GROUP for class: org.gradle.api.publication.maven.internal.ant.DefaultGroovyMavenDeployer

I guess this might have something to do with gradle.properties or local.properties but I am not sure so I tried to copy those from other projects but it still failed, so any workarounds that you might have in your mind to fix this

**Pull request merged**